### ISSUE TITLE
Totals Information Management - setting quote shipping method when address method is null

### ISSUE THREAD

**Issue by joe-vortex** [magento/magento2#25147]
### Preconditions (*)

1. 2.3.2(2.3-develop) Magento Commerce or Open Source 
2. Amasty Conditions module installed (cannot reproduce with vanilla m2 because this)
3. PHP 7.2.21

### Steps to reproduce (*)

1. On the checkout, the component `Amasty_Conditions/js/action/recollect-totals` makes a call to `/V1/guest-carts/:cartId/totals-information` which passes through Magento\Checkout\Model\TotalsInformationManagement
2. In this model the following:
```
$quote->getShippingAddress()->setCollectShippingRates(true)->setShippingMethod(
     $addressInformation->getShippingCarrierCode() . '_' . $addressInformation->getShippingMethodCode()
);
``` 
is set, but when a shipping method has not been selected, it will return the quote shipping method with a total of 0 and a name of `_`, because `$addressInformation->getShippingCarrierCode()` and method code are null - so `setShippingMethod('_')` happens in effect.
3. This then means that validation for the shipping method is bypassed.
4. An order can then be placed without shipping information/carrier

### Expected result (*)

1. If the address info has the shipping carrier and/or method code set to null, the `setShippingMethod` should not be called and the quote total is returned without setting a shipping code pair

### Actual result (*)

1. When a address info instance is submitted with shipping codes set to null, a method is set with label `_` and zero value, meaning an order can be placed without a valid shipping method.

**Comment by joe-vortex** [comment]
can be fixed with the following patch:
```
--- magento/module-checkout/Model/TotalsInformationManagement.php	2019-10-18 13:17:40.275177439 +0100
+++ magento/module-checkout/Model/TotalsInformationManagement.php	2019-10-18 13:17:30.307000000 +0100
@@ -52,9 +52,11 @@
             $quote->setBillingAddress($addressInformation->getAddress());
         } else {
             $quote->setShippingAddress($addressInformation->getAddress());
-            $quote->getShippingAddress()->setCollectShippingRates(true)->setShippingMethod(
-                $addressInformation->getShippingCarrierCode() . '_' . $addressInformation->getShippingMethodCode()
-            );
+            if ($addressInformation->getShippingCarrierCode() && $addressInformation->getShippingMethodCode()) {
+                $quote->getShippingAddress()->setCollectShippingRates(true)->setShippingMethod(
+                    $addressInformation->getShippingCarrierCode().'_'.$addressInformation->getShippingMethodCode()
+                );
+            }
         }
         $quote->collectTotals();

```

**Comment by sdzhepa** [comment]
After discussion and investigation, I believe this issue is not specific only for Magento Commerce(EE) and the same problem should be in Open Source(CE) as well. It could be not possible to reproduce it by provided steps without buying the `Amasty Conditions module` but from a code perspective, we see the bug.

Confirmed

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @sdzhepa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-22098` were created

**Issue Available**: @sdzhepa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by joe-vortex** [comment]
thanks @sdzhepa , are you happy for me to make a PR?

**Comment by magento-engcom-team** [comment]
Hi @joe-vortex. Thank you for your report. 
The issue has been fixed in magento/magento2#25510 by @joe-vortex in 2.4-develop branch
Related commit(s):
 - [a99d980cebbfa1495de25aa91b9c70d86b2da7a6](https://github.com/magento/magento2/commit/a99d980cebbfa1495de25aa91b9c70d86b2da7a6)
 - [1900ffcddcd8e9d3817de7fea29496d69997d531](https://github.com/magento/magento2/commit/1900ffcddcd8e9d3817de7fea29496d69997d531)
 - [8820a5c45a8652b48e1875933629fea2717c2e56](https://github.com/magento/magento2/commit/8820a5c45a8652b48e1875933629fea2717c2e56)
 - [b0e0517f8a570d9b98312f472ed147ab3b7331f3](https://github.com/magento/magento2/commit/b0e0517f8a570d9b98312f472ed147ab3b7331f3)
 - [e4fb9d153ef4ea1c411de5ac7954a4849e57a7f0](https://github.com/magento/magento2/commit/e4fb9d153ef4ea1c411de5ac7954a4849e57a7f0)

The fix will be available with the upcoming 2.4.2 release.
